File: python/base_engine.py
"""
base_engine.py
Shared file-matching logic inherited by MBEngine and SpotifyEngine.
Both engines expose the same interface so music_organizer can swap them freely.
"""

import os
import re
from thefuzz import process, fuzz

# Minimum fuzzy score (0–100) to accept a match.
# Lower = more permissive, higher = stricter.
TRACK_MATCH_THRESHOLD  = 30   # file → track name match
ALBUM_MATCH_THRESHOLD  = 80   # folder → release title match (batch mode)


class BaseEngine:

    # ── Filename sanitizer ────────────────────────────────────────────────────

    def _normalize_for_matching(self, title):
        """
        Normalise a title for fuzzy comparison only — never used for output names.
        Handles all common 'featuring' formats and strips 'The' prefixes.
        """
        s = title.lower()
        s = re.sub(r"[\(\[\{]", " ", s)
        s = re.sub(r"[\)\]\}]", " ", s)
        s = re.sub(r"\b(featuring|feat\.?|ft\.?|with|w/)\s+", " feat ", s)
        s = re.sub(r"\bthe\s+", " ", s)
        s = re.sub(r"\s+", " ", s).strip()
        return s

    def _subset_sort_ratio(self, s1, s2):
        """
        token_sort_ratio plus a subset boost:
        if ALL of s1's tokens appear in s2 (and s1 has ≥ 2 tokens), boost to
        at least 85. This handles files named with just the song title when
        Spotify appends '(with X feat. Y)' or version tags.

        Minimum 2 tokens prevents single-word titles ('War', 'Zombie') from
        accidentally hitting longer tracks they don't belong to.
        """
        base      = fuzz.token_sort_ratio(s1, s2)
        fn_tokens = set(s1.split())
        tr_tokens = set(s2.split())
        if len(fn_tokens) >= 2 and fn_tokens.issubset(tr_tokens):
            base = max(base, 85)
        return base
        """
        Normalise a title for fuzzy comparison only — never used for output names.

        Handles all common 'featuring' formats:
          ft. / feat. / featuring / with / w/ / (ft. …) / [feat. …]
        Strips 'The ' prefix from artist names so 'The Smashing Pumpkins'
        matches 'Smashing Pumpkins'.
        Lowercases and collapses whitespace.
        """
        s = title.lower()

        # Normalise all featuring variations to a plain space-separated token
        # so "feat.", "ft.", "featuring", "w/", "with" all look the same
        s = re.sub(r"[\(\[\{]", " ", s)
        s = re.sub(r"[\)\]\}]", " ", s)
        s = re.sub(
            r"\b(featuring|feat\.?|ft\.?|with|w/)\s+",
            " feat ", s
        )

        # Strip leading 'the ' from each artist name after feat marker
        # e.g. "feat the smashing pumpkins" → "feat smashing pumpkins"
        s = re.sub(r"\bthe\s+", " ", s)

        # Collapse whitespace
        s = re.sub(r"\s+", " ", s).strip()
        return s

    def _sanitize(self, title):
        """Replace illegal filename characters with a space, collapse doubles."""
        cleaned = "".join(c if c not in r'\/*?:"<>|' else " " for c in title)
        return re.sub(r" {2,}", " ", cleaned).strip()

    _FEAT_IN_TITLE = re.compile(
        r'\(feat\.|\(ft\.|\(with\b|\bfeat\.|\bft\.|\bfeaturing\b', re.I
    )

    def _track_filename(self, track, ext):
        """
        Append featured artists only when the Spotify title doesn't already
        embed featuring info. If the title contains any feat/ft/with marker
        the Spotify data already has the full info — don't double-append
        (catches alias mismatches like 'Machine Gun Kelly' vs 'MGK').
        """
        name     = self._sanitize(track["name"])
        featured = track.get("featured", [])
        if featured and not self._FEAT_IN_TITLE.search(track["name"]):
            title_lc  = track["name"].lower()
            new_feats = [a for a in featured if a.lower() not in title_lc]
            if new_feats:
                feat_str = " & ".join(self._sanitize(a) for a in new_feats)
                name = f"{name} ft. {feat_str}"
        return f"{name}{ext}"

    # ── Single-album preview ──────────────────────────────────────────────────

    def generate_preview(self, folder_path, tracks, f_offset=0):
        """
        NUMERIC: Match files (sorted alphabetically) to tracks in order.
        No track-number prefix on output filenames.
        Returns ([(old, new, track_num, disc_num), …], log_string).
        """
        try:
            files = sorted([
                f for f in os.listdir(folder_path)
                if os.path.isfile(os.path.join(folder_path, f))
            ])
        except Exception as e:
            return [], f"Error reading folder: {e}"

        ep_data, log = [], "NUMERIC MATCH:\n"
        for i, f in enumerate(files):
            idx = i + f_offset
            if 0 <= idx < len(tracks):
                ext      = os.path.splitext(f)[1]
                new_name = self._track_filename(tracks[idx], ext)
                ep_data.append((f, new_name, tracks[idx]["num"], tracks[idx].get("disc", 1), 100))
                log += f"  {f}\n  → {new_name}\n\n"
        return ep_data, log

    def generate_fuzzy_preview(self, folder_path, tracks, f_offset=0):
        """
        FUZZY: Match filenames to track names using token_sort_ratio (respects
        full track title including suffixes like Pt. II, feat., etc.).
        Strips leading track numbers and common scene tags before matching.
        Guards against duplicate output filenames by appending disc/track info.
        Returns ([(old, new, track_num, disc_num), …], log_string).
        """
        try:
            files = [
                f for f in os.listdir(folder_path)
                if os.path.isfile(os.path.join(folder_path, f))
            ]
        except Exception as e:
            return [], f"Error: {e}"

        track_map    = {t["name"]: t for t in tracks}
        track_titles = list(track_map.keys())

        # ── Detect ambiguous track names ──────────────────────────────────────
        # Count from the RAW tracks list (not track_map which deduplicates by name)
        # so that e.g. "War" on disc 1 AND disc 2 both get counted.
        from collections import Counter
        norm_counts     = Counter(self._normalize_for_matching(t["name"]) for t in tracks)
        ambiguous_norms = {n for n, c in norm_counts.items() if c > 1}

        # Pre-build normalized → original title map
        norm_map    = {self._normalize_for_matching(t): t for t in track_titles}
        norm_titles = list(norm_map.keys())

        temp_matches, log = [], "FUZZY MATCH:\n"
        used_tracks  = set()   # track names already claimed — one file per track

        for f in files:
            fn_base = os.path.splitext(f)[0].replace(".", " ").replace("_", " ")
            fn_base = re.sub(
                r"(1080p|720p|x264|x265|HEVC|WEB-DL|BluRay|\d{3}kbps)",
                "", fn_base, flags=re.I
            )
            fn_base = re.sub(r"^\d+\s*[-\.]\s*", "", fn_base).strip()
            fn_norm = self._normalize_for_matching(fn_base)

            remaining_norm = [t for t in norm_titles
                              if norm_map[t] not in used_tracks]
            if not remaining_norm:
                log += f"  SKIPPED (no tracks left): {f}\n"
                continue

            match = process.extractOne(fn_norm, remaining_norm,
                                       scorer=self._subset_sort_ratio)
            if match and match[1] > TRACK_MATCH_THRESHOLD:
                matched_norm   = match[0]
                original_title = norm_map[matched_norm]

                # ── Ambiguity guard ───────────────────────────────────────────
                # If this normalized name maps to multiple tracks (e.g. "War"
                # on disc 1 and disc 2), skip auto-matching and let the user
                # decide manually in the pairing dialog.
                if matched_norm in ambiguous_norms:
                    log += f"  AMBIGUOUS: {f}  (multiple tracks share this name — use manual pairing)\n"
                    continue

                track = track_map[original_title]
                used_tracks.add(original_title)
                temp_matches.append({
                    "old_name": f,
                    "num":      track["num"],
                    "disc":     track.get("disc", 1),
                    "title":    track["name"],
                    "track":    track,          # full track object for featured artists
                    "score":    match[1],
                })
            else:
                log += f"  UNMATCHED: {f}  (see manual pairing dialog)\n"

        temp_matches.sort(key=lambda x: (x["disc"], x["num"]))

        # ── Duplicate output-name guard ───────────────────────────────────────
        # If two matches would produce the same filename (e.g. disc 1 "Zombie"
        # and disc 2 "Zombie (feat. Smashing Pumpkins)" both sanitize to
        # "Zombie.flac"), disambiguate by prefixing with disc or track number.
        seen_names = {}   # sanitized_name → first match index
        for m in temp_matches:
            key = self._sanitize(m["title"]).lower()
            if key in seen_names:
                # Mark both the earlier and current match as needing disambiguation
                seen_names[key].append(m)
            else:
                seen_names[key] = [m]

        needs_disambig = {
            id(m)
            for matches in seen_names.values() if len(matches) > 1
            for m in matches
        }

        ep_data = []
        for m in temp_matches:
            ext      = os.path.splitext(m["old_name"])[1]
            base_name = self._track_filename(m["track"], ext)

            if id(m) in needs_disambig:
                prefix   = (f"Disc {m['disc']} - " if m["disc"] > 1
                            else f"{m['num']:02d} - ")
                stem     = os.path.splitext(base_name)[0]
                new_name = f"{prefix}{stem}{ext}"
                note     = f" [disambiguated: {prefix.strip()}]"
            else:
                new_name = base_name
                note     = ""

            ep_data.append((m["old_name"], new_name, m["num"], m["disc"], m["score"]))
            log += f"  [{m['score']}%] {m['old_name']}\n  → {new_name}{note}\n\n"

        return ep_data, log

    # ── Batch preview ─────────────────────────────────────────────────────────

    def generate_batch_preview(self, artist_path, artist_id,
                               match_mode="Numeric", f_offset=0):
        """
        BATCH: Walk every album subfolder under artist_path, fuzzy-match each
        to a release from get_releases(), fetch its tracks, and build preview pairs.

        Returns a list of dicts:
            {folder, path, release (or None), score, ep_data, log}
        """
        try:
            album_folders = sorted([
                d for d in os.listdir(artist_path)
                if os.path.isdir(os.path.join(artist_path, d))
            ])
        except Exception:
            return []

        releases = self.get_releases(artist_id)
        if not releases:
            return []

        release_map    = {r["title"]: r for r in releases}
        release_titles = list(release_map.keys())
        results        = []

        for folder in album_folders:
            album_path = os.path.join(artist_path, folder)
            # Strip leading year / number prefix for cleaner matching
            clean_folder = re.sub(r"^\d{4}\s*[-\.]\s*", "", folder).strip()

            match = process.extractOne(
                clean_folder, release_titles, scorer=fuzz.token_set_ratio
            )
            score = match[1] if match else 0

            if match and score >= ALBUM_MATCH_THRESHOLD:
                release = release_map[match[0]]
                tracks  = self.get_tracks(release["id"])

                if tracks:
                    if match_mode == "Numeric":
                        ep_data, log = self.generate_preview(album_path, tracks, f_offset)
                    else:
                        ep_data, log = self.generate_fuzzy_preview(album_path, tracks, f_offset)

                    header = (
                        f"📁 {folder}\n"
                        f"   ↳ {release['title']} ({release['year']})  [{score}% match]\n"
                        f"   {len(ep_data)} files matched\n"
                    )
                    results.append({
                        "folder":  folder,
                        "path":    album_path,
                        "release": release,
                        "score":   score,
                        "ep_data": ep_data,
                        "log":     header + log,
                    })
                else:
                    results.append({
                        "folder": folder, "path": album_path,
                        "release": release, "score": score,
                        "ep_data": [],
                        "log": f"📁 {folder}\n   ↳ {release['title']} — no tracks found.\n",
                    })
            else:
                results.append({
                    "folder": folder, "path": album_path,
                    "release": None, "score": score,
                    "ep_data": [],
                    "log": f"📁 {folder}\n   ✗ No match found (best score: {score}%)\n",
                })

        return results

File: c++ source code/base_engine.cpp
#include "base_engine.h"
#include "fuzzy.h"
#include <algorithm>
#include <filesystem>
#include <regex>
#include <sstream>
#include <set>
#include <map>
#include <unordered_map>

namespace fs = std::filesystem;

// ── Normalise for matching ────────────────────────────────────────────────────

std::string BaseEngine::normalizeForMatching(const std::string& title) const {
    std::string s = title;
    // Lowercase
    std::transform(s.begin(), s.end(), s.begin(), ::tolower);
    // Open/close brackets → space
    s = std::regex_replace(s, std::regex(R"([\(\[\{])"), " ");
    s = std::regex_replace(s, std::regex(R"([\)\]\}])"), " ");
    // Normalise featuring variants
    s = std::regex_replace(s,
        std::regex(R"(\b(featuring|feat\.?|ft\.?|with|w/)\s+)",
                   std::regex_constants::icase),
        " feat ");
    // Strip leading "the "
    s = std::regex_replace(s, std::regex(R"(\bthe\s+)"), " ");
    // Collapse whitespace
    s = std::regex_replace(s, std::regex(R"(\s+)"), " ");
    // Trim
    size_t start = s.find_first_not_of(' ');
    size_t end   = s.find_last_not_of(' ');
    if (start == std::string::npos) return "";
    return s.substr(start, end - start + 1);
}

// ── Subset sort ratio ─────────────────────────────────────────────────────────

int BaseEngine::subsetSortRatio(const std::string& s1, const std::string& s2) const {
    return fuzzy::subset_sort_ratio(s1, s2);
}

// ── Sanitize (illegal filesystem chars) ──────────────────────────────────────

std::string BaseEngine::sanitize(const std::string& title) const {
    static const std::string illegal = R"(\/*?:"<>|)";
    std::string out;
    out.reserve(title.size());
    for (char c : title)
        out += (illegal.find(c) != std::string::npos) ? ' ' : c;
    // Collapse double spaces
    out = std::regex_replace(out, std::regex(R"( {2,})"), " ");
    // Trim
    size_t s = out.find_first_not_of(' ');
    size_t e = out.find_last_not_of(' ');
    if (s == std::string::npos) return "";
    return out.substr(s, e - s + 1);
}

// ── Track filename ────────────────────────────────────────────────────────────

std::string BaseEngine::trackFilename(const Track& track, const std::string& ext) const {
    static const std::regex feat_in_title(
        R"(\(feat\.|\(ft\.|\(with\b|\bfeat\.|\bft\.|\bfeaturing\b)",
        std::regex_constants::icase);

    std::string name = sanitize(track.name);

    if (!track.featured.empty() &&
        !std::regex_search(track.name, feat_in_title))
    {
        // Only append artists not already mentioned in the title
        std::string title_lc = track.name;
        std::transform(title_lc.begin(), title_lc.end(), title_lc.begin(), ::tolower);

        std::vector<std::string> new_feats;
        for (auto& a : track.featured) {
            std::string a_lc = a;
            std::transform(a_lc.begin(), a_lc.end(), a_lc.begin(), ::tolower);
            if (title_lc.find(a_lc) == std::string::npos)
                new_feats.push_back(a);
        }
        if (!new_feats.empty()) {
            std::string feat_str;
            for (size_t i = 0; i < new_feats.size(); ++i) {
                if (i) feat_str += " & ";
                feat_str += sanitize(new_feats[i]);
            }
            name += " ft. " + feat_str;
        }
    }
    return name + ext;
}

// ── Numeric preview ───────────────────────────────────────────────────────────

std::pair<std::vector<MatchResult>, std::string>
BaseEngine::generatePreview(const std::string& folderPath,
                            const std::vector<Track>& tracks,
                            int fOffset)
{
    std::vector<std::string> files;
    try {
        for (auto& entry : fs::directory_iterator(folderPath))
            if (entry.is_regular_file())
                files.push_back(entry.path().filename().string());
    } catch (...) {
        return {{}, "Error reading folder.\n"};
    }
    std::sort(files.begin(), files.end());

    std::vector<MatchResult> results;
    std::string log = "NUMERIC MATCH:\n";

    for (int i = 0; i < (int)files.size(); ++i) {
        int idx = i + fOffset;
        if (idx < 0 || idx >= (int)tracks.size()) continue;
        const auto& t = tracks[idx];
        std::string ext = fs::path(files[i]).extension().string();
        std::string newName = trackFilename(t, ext);
        results.push_back({files[i], newName, t.num, t.disc, 100});
        log += "  " + files[i] + "\n  → " + newName + "\n\n";
    }
    return {results, log};
}

// ── Fuzzy preview ─────────────────────────────────────────────────────────────

std::pair<std::vector<MatchResult>, std::string>
BaseEngine::generateFuzzyPreview(const std::string& folderPath,
                                 const std::vector<Track>& tracks,
                                 int fOffset)
{
    (void)fOffset;
    std::vector<std::string> files;
    try {
        for (auto& entry : fs::directory_iterator(folderPath))
            if (entry.is_regular_file())
                files.push_back(entry.path().filename().string());
    } catch (...) {
        return {{}, "Error reading folder.\n"};
    }

    // Build track maps
    std::map<std::string, const Track*> track_map;   // name → track
    std::vector<std::string> track_titles;
    for (auto& t : tracks) {
        if (track_map.find(t.name) == track_map.end()) {
            track_map[t.name] = &t;
            track_titles.push_back(t.name);
        }
    }

    // Count normalised track names to detect ambiguity (e.g. "War" on disc 1 and 2)
    std::map<std::string, int> norm_counts;
    for (auto& t : tracks)
        norm_counts[normalizeForMatching(t.name)]++;
    std::set<std::string> ambiguous_norms;
    for (auto& [n, c] : norm_counts)
        if (c > 1) ambiguous_norms.insert(n);

    // Pre-build normalised → original title map
    std::map<std::string, std::string> norm_map;
    std::vector<std::string> norm_titles;
    for (auto& title : track_titles) {
        std::string n = normalizeForMatching(title);
        norm_map[n] = title;
        norm_titles.push_back(n);
    }

    struct TempMatch {
        std::string old_name;
        int num, disc;
        std::string title;
        const Track* track;
        int score;
    };

    std::vector<TempMatch> temp_matches;
    std::set<std::string> used_tracks;
    std::string log = "FUZZY MATCH:\n";

    static const std::regex scene_tags(
        R"((1080p|720p|x264|x265|HEVC|WEB-DL|BluRay|\d{3}kbps))",
        std::regex_constants::icase);
    static const std::regex leading_num(R"(^\d+\s*[-\.]\s*)");

    for (auto& fname : files) {
        std::string fn_base = fs::path(fname).stem().string();
        // Replace . and _ with space
        std::replace(fn_base.begin(), fn_base.end(), '.', ' ');
        std::replace(fn_base.begin(), fn_base.end(), '_', ' ');
        fn_base = std::regex_replace(fn_base, scene_tags, "");
        fn_base = std::regex_replace(fn_base, leading_num, "");
        // Trim
        size_t s = fn_base.find_first_not_of(' ');
        if (s != std::string::npos) fn_base = fn_base.substr(s);

        std::string fn_norm = normalizeForMatching(fn_base);

        // Only consider unclaimed tracks
        std::vector<std::string> remaining;
        for (auto& n : norm_titles)
            if (used_tracks.find(norm_map.at(n)) == used_tracks.end())
                remaining.push_back(n);

        if (remaining.empty()) {
            log += "  SKIPPED (no tracks left): " + fname + "\n";
            continue;
        }

        auto scorer = [this](const std::string& a, const std::string& b) {
            return this->subsetSortRatio(a, b);
        };
        auto [best_norm, best_score] = fuzzy::extract_one(fn_norm, remaining, scorer);

        if (best_score > TRACK_MATCH_THRESHOLD) {
            if (ambiguous_norms.count(best_norm)) {
                log += "  AMBIGUOUS: " + fname +
                       "  (multiple tracks share this name — use manual pairing)\n";
                continue;
            }
            std::string orig_title = norm_map.at(best_norm);
            const Track* track = track_map.at(orig_title);
            used_tracks.insert(orig_title);
            temp_matches.push_back({fname, track->num, track->disc, track->name,
                                    track, best_score});
        } else {
            log += "  UNMATCHED: " + fname + "  (see manual pairing dialog)\n";
        }
    }

    // Sort by disc then track number
    std::sort(temp_matches.begin(), temp_matches.end(),
              [](const TempMatch& a, const TempMatch& b) {
                  if (a.disc != b.disc) return a.disc < b.disc;
                  return a.num < b.num;
              });

    // Duplicate output-name guard
    std::map<std::string, std::vector<int>> seen_names; // sanitized title → indices
    for (int i = 0; i < (int)temp_matches.size(); ++i) {
        std::string key = sanitize(temp_matches[i].title);
        std::transform(key.begin(), key.end(), key.begin(), ::tolower);
        seen_names[key].push_back(i);
    }
    std::set<int> needs_disambig;
    for (auto& [k, idxs] : seen_names)
        if (idxs.size() > 1)
            for (int i : idxs) needs_disambig.insert(i);

    std::vector<MatchResult> ep_data;
    for (int i = 0; i < (int)temp_matches.size(); ++i) {
        auto& m = temp_matches[i];
        std::string ext     = fs::path(m.old_name).extension().string();
        std::string base_nm = trackFilename(*m.track, ext);
        std::string new_name;
        std::string note;

        if (needs_disambig.count(i)) {
            std::string prefix = (m.disc > 1)
                ? "Disc " + std::to_string(m.disc) + " - "
                : (m.num < 10 ? "0" : "") + std::to_string(m.num) + " - ";
            std::string stem = fs::path(base_nm).stem().string();
            new_name = prefix + stem + ext;
            note = " [disambiguated: " + prefix + "]";
        } else {
            new_name = base_nm;
        }
        ep_data.push_back({m.old_name, new_name, m.num, m.disc, m.score});
        log += "  [" + std::to_string(m.score) + "%] " + m.old_name +
               "\n  → " + new_name + note + "\n\n";
    }
    return {ep_data, log};
}

// ── Batch preview ─────────────────────────────────────────────────────────────

std::vector<BatchResult>
BaseEngine::generateBatchPreview(const std::string& artistPath,
                                 const std::string& artistId,
                                 const std::string& matchMode,
                                 int fOffset)
{
    std::vector<std::string> folders;
    try {
        for (auto& entry : fs::directory_iterator(artistPath))
            if (entry.is_directory())
                folders.push_back(entry.path().filename().string());
    } catch (...) {
        return {};
    }
    std::sort(folders.begin(), folders.end());

    auto releases = getReleases(artistId);
    if (releases.empty()) return {};

    std::map<std::string, const Release*> release_map;
    std::vector<std::string> release_titles;
    for (auto& r : releases) {
        release_map[r.title] = &r;
        release_titles.push_back(r.title);
    }

    static const std::regex year_prefix(R"(^\d{4}\s*[-\.]\s*)");
    std::vector<BatchResult> results;

    for (auto& folder : folders) {
        std::string album_path = artistPath + "/" + folder;
        std::string clean = std::regex_replace(folder, year_prefix, "");
        // Trim
        size_t s = clean.find_first_not_of(' ');
        if (s != std::string::npos) clean = clean.substr(s);

        auto [best_title, score] = fuzzy::extract_one(clean, release_titles,
                                                       fuzzy::token_set_ratio);

        if (!best_title.empty() && score >= ALBUM_MATCH_THRESHOLD) {
            const Release* rel = release_map.at(best_title);
            auto tracks = getTracks(rel->id);
            if (!tracks.empty()) {
                std::pair<std::vector<MatchResult>, std::string> pv;
                if (matchMode == "Numeric")
                    pv = generatePreview(album_path, tracks, fOffset);
                else
                    pv = generateFuzzyPreview(album_path, tracks, fOffset);

                std::string header =
                    "📁 " + folder + "\n"
                    "   ↳ " + rel->title + " (" + rel->year + ")  [" +
                    std::to_string(score) + "% match]\n"
                    "   " + std::to_string(pv.first.size()) + " files matched\n";

                results.push_back({folder, album_path, true, *rel, score,
                                   pv.first, header + pv.second});
            } else {
                results.push_back({folder, album_path, true, *rel, score, {},
                    "📁 " + folder + "\n   ↳ " + rel->title + " — no tracks found.\n"});
            }
        } else {
            results.push_back({folder, album_path, false, {}, score, {},
                "📁 " + folder + "\n   ✗ No match found (best score: " +
                std::to_string(score) + "%)\n"});
        }
    }
    return results;
}
